### DISCUSSION TITLE
How to get the current target name in parcel transformer?

### DISCUSSION THREAD

**Discussion by ghost** [parcel-bundler/parcel#7158] [Q&A] [answered]
Let's assume that our `package.json` contains two targets named `server` and `client` (names chosen arbitrarily):

```json
    "targets": {
        "server": {
            "source": "src/main.js",
            "context": "node",
            ...
        },
        "client": {
            "source": "src/script.js",
            "context": "browser",
            ...
        }
    }
```

And we use a no-op `*.js` transformer like the one that parcel 2 documentation suggests:

```javascript
import {Transformer} from '@parcel/plugin';

export default new Transformer({
  async transform({asset}) {
    let source = await asset.getCode();
    let sourceMap = await asset.getMap();

    // asset.env.context is either 'node' or 'browser'
    // but how to get target name, i.e. 'server' or 'client'?

    asset.setCode(source);
    asset.setMap(sourceMap);

    return [asset];
  }
});
```

**How to obtain the current target name (`'server'` or `'client'`) in this transformer?**

Things I've tried so far:
- searching for "[parcel] transformer target [name]" in issues, discussions and google;
- passing the custom `targetname` option around `source`/`context` in `package.json`;
- investigated what you can get from `asset` fields/methods and `transform` method arguments, recursively;
- studied all references to the `Target` interface in parcel's type definitions;
- looked up for some correlation between asset/context/etc `id`s and package.json fields.

Neither of these worked for me.

Things that I don't want:
- figuring out the target name from `asset.env.context`, because there may be multiple targets for e.g. `node` context;
- figuring it out by the `asset.filePath`, because transformer also works on `import`ed sources, which could also be imported into both targets (e.g. `src/client-server-shared.js`).

Thanks!

**Comment by mischnic** [comment, marked as answer]
As you've already noticed, this isn't possible by design because two targets can share assets (which are then only transformed once). So there isn't always a single "current" target for a specific asset.

The only information available coming from the target is in `asset.env`.

**Reply by ghost** [reply]
Thank you for the quick answer! So, if I understand it correctly, if I had 3 entry points `src/main.js` (node), `src/script.js` (browser) and `src/migrate.js` (node), all of which import some `src/shared.js`, then that `shared.js` would be transformed only 2 times instead of 3, because it only participates in two contexts: node and browser. Is that correct?
